### ISSUE TITLE
Order of operations in rendering a frame

### ISSUE THREAD

**Issue by Yamakaky** [gfx-rs/wgpu-rs#319]
I have some data coming from a sensor at a variable rate independent from the refresh rate. I want to:
- Copy the data to a buffer
- Use a compute shader to transform this data into vertices with position and normal
- Use these vertices as input for a render pipeline.

Each step must be completed in order, and the first two only run when I get new data. I also want the render pipeline to see a fully computed buffer (similar to tearing problems). As I see it, there is multiple ways to do it (including nonsensical ones):
- Use a different encoder and render/compute pass per step
- Use an encoder for the first two with two passes and another encoder for the render
- Use one encoder for everything

What are the ordering guarantees of each solution and which one should I use? My intuition is that operations from the same encoder are ordered as defined in the code and the queue can reorder/overlap command buffers, so would need to use a single encoder, is that correct? What about if the compute shader takes multiple frames to finish? I don't think there is a way to define fences where I would have two vertex buffers and atomically switch between the two when the compute shader finishes?

I plan do improve wgpu-rs doc in the following days, so my question is a bit more general that my personal use case.

**Comment by kvark** [comment]
The queue doesn't re-order or overlap command buffers. It behaves as if everything is executed sequentially. We guarantee that if resource is used in encoder A, then encoder B, then another command buffer C (upon submission), then access to it is properly synchronized.

TL;DR: any of the three choices you mentioned are fine. If I were doing this, I'd just submit a separate command buffer to process sensor data. This would mean GPU can work on it while you are waiting for Vsync for the next frame.

**Comment by Yamakaky** [comment]
Hum OK, then the command buffers are just a performance optimization and have the same semantic as pushing operations one by one on the queue?

**Comment by kvark** [comment]
Semantically, yes.
In terms of performance there is a CPU cost to call `submit` on a queue, but it can allow the GPU to start working on the operations sooner.

**Comment by Yamakaky** [comment]
OK, thanks!
